enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,585 @@
+---
+title: add-ImagePolicy-to-config.openshift.io-v1alpha1
+authors:
+  - "@QiWang19"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@saschagrunert, for API design, implementation details and graduation criteria"
+  - "@mrunalp"
+  - "@yuqi-zhang"
+  - "@mtrmac"
+  - "@wking"
+  - "@ingvagabund"
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - "@mrunalp"
+  - "@JoelSpeed"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@mrunalp"
+  - "@JoelSpeed"
+creation-date: 2023-05-17
+last-updated: 2023-06-22
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/OCPNODE-1628
+---
+
+# Add CRD ImagePolicy to config.openshift.io/v1alpha1
+
+## Summary
+
+This enhancement propose an ImagePolicy CRD to provide a way to set sigstore image signature verification configuration on Openshift.
+
+## Motivation
+
+### User Stories
+
+As an OpenShift user, I want to verify the container images signed using sigstore
+tools, so that I can utilize the increased security of my software supply chain.
+
+### Goals
+
+- The user can create an ImagePolicy instance specifying the images/repositories to be verified and their policy. The MCO will write the configuration for signature verification. Once this is done, CRI-O can verify the images/repositories.
+
+- MCO container runtime config controller watches ImagePolicy instance in different kubernetes namespaces and merges the instances for each namespace into a single [containers-policy.json](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md) in a predefined `/path/to/policies/\<NAMESPACE\>.json`.
```

Comment:
the policy scoped to the namespace. https://github.com/openshift/enhancements/pull/1402/files#diff-688e8f3b56284c3f0e907836b95cbcb60bcd0676c4819a2e4b1524529bf8d5c5R43 mentioned the CRIO will choose the specific <namespace>.json using the namespace from the sandbox config.

could you specify what kind of conflict? The [Merge rules](https://github.com/openshift/enhancements/pull/1402/files#diff-688e8f3b56284c3f0e907836b95cbcb60bcd0676c4819a2e4b1524529bf8d5c5R264) has explained the policies on same namespaces.

Reply:
What would happen if I put two different policies for the same image scope for example? This is defining how the images should be signed in a particular repository right? But, what happens if i have two different configs for how an image should be signed? In that case, which one would it chose to apply?

Reply:
That's documented at [policy.json](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md#policy-requirements)

If an image has two different policies configured at the following level respectively:
- (image level, repo level) : If multiple policy requirements match a given image, only the requirements from the most specific match apply, the more general policy requirements definitions are ignored.

- (image level, image level), (repo level, repo level): If multiple policy requirements match a given image, only the requirements from the most specific match apply, the more general policy requirements definitions are ignored.

Reply:
That actually seems to highlight a possible gap:

Currently, the merge rules from this PR mean that if the global `openshift-config` and an unprivileged per-namespace CR specify policy for _exactly the same scope_, that conflict is resolved by ignoring the unprivileged per-namespace policy, and using the global `openshift-config` policy; with the intent (IIRC) to let a cluster admin define mandatory policies.

But with the `policy.json` semantics, unprivileged namespaces could avoid those mandatory policies by using more narrow scopes (e.g. a scope for every image in a repo, instead of a scope for the whole repo).

My first instinct is to say this should be prohibited, i.e. unprivileged per-namespace scopes nested _inside_ global `openshift-config` defined scopes should be ignored/rejected, the same way the exactly-equal scopes are ignored/rejected.

Admittedly, I’m not as confident about the rejection of nested scopes as I’m about the rejection of exactly-equal scopes: with a very broad cluster-wide scope (e.g. for all of `company-registry.example.com`) defined on the cluster level, there might(?) be legitimate use cases for letting unprivileged users define other policies for individual repos, or at least for letting unprivileged users impose extra restrictions on top of the cluster-wide policy.

On balance, I think it would be better to reject per-namespace scopes nested inside mandatory cluster-wide scopes, in the first version. I don’t think it’s _too likely_ that users will run into this situation, and if they do, we can think about relaxing this (adding a field to cluster-wide scopes, to let them opt in into being overridden by per-namespace scopes?) in the future.

Reply:
It strikes me that there are a lot of decisions and scenarios that need to be played through about the permissions and potential ways a namespace scope could override the global scope. Is it documented somewhere all of the overlap, including these latest comments?

Perhaps some examples would help or a table that explains the behaviour?